fabtests/hpcs/src/test/sendrecv.c

```diff
@@ -0,0 +1,240 @@
+		{"workqueue", no_argument, 0, 'w'},
+		{"help", no_argument, 0, 'h'},
+		{0}
+	};
+
+	struct test_arguments *args = calloc(sizeof(struct test_arguments), 1);
+	if (args == NULL)
+		return -ENOMEM;
+
+	*args = (struct test_arguments) {
+		.transfer_size = DEFAULT_TRANSFER_SIZE
+	};
+
+	while ((op = getopt_long(argc, argv, "s:wh", longopt, &longopt_idx)) != -1) {
+		switch (op) {
+		case 's':
+			if (sscanf(optarg, "%zu", &args->transfer_size) != 1)
+			return -EINVAL;
+			break;
+		case 'w':
+			args->use_workqueue = 1;
+			break;
+		case 'h':
+		default:
+			fprintf(stderr, "<test arguments> :=\n"
+					"\t[-s | --size=<size>]\n"
+					"\t[-w | --workqueue]\n"
+					"\t[-h | --help]\n");
+			return -EINVAL;
+		}
+	}
+
+	*buffer_size = args->transfer_size;
+	*arguments = args;
+
+	return 0;
+}
+
+void free_arguments (struct test_arguments *arguments)
+{
+	free (arguments);
+	return;
+}
+
+static struct test_config config(const struct test_arguments *arguments)
+{
+	struct test_config config = {
+		.minimum_caps = FI_TAGGED | FI_SEND | FI_RECV,
+
+		.tx_use_cntr = true,
+		.rx_use_cntr = true,
+		.tx_use_cq = true,
+		.rx_use_cq = true,
+
+		.rx_use_mr = false,
+
+		.tx_buffer_size = arguments->transfer_size,
+		.rx_buffer_size = arguments->transfer_size,
+		.tx_buffer_alignment = 0,
+		.rx_buffer_alignment = 0,
+
+		.tx_data_object_sharing = DATA_OBJECT_PER_DOMAIN,
+		.rx_data_object_sharing = DATA_OBJECT_PER_TRANSFER,
+
+		.tx_context_count = 1,
+		.rx_context_count = 1,
+	};
+
+	return config;
+}
+
+static int tx_transfer(const struct test_arguments *arguments,
+		const size_t transfer_id,
+		const size_t transfer_count,
+		const fi_addr_t rx_address,
+		struct fid_ep *endpoint,
+		struct op_context *op_context,
+		uint8_t *buffer,
+		void *desc,
+		uint64_t key,
+		int rank,
+		uint64_t flags)
+{
+	const uint64_t tag = transfer_id;
+	struct context_info *ctxinfo = op_context->ctxinfo;
+	int ret;
+
+	ctxinfo->iov = (struct iovec) {
+		.iov_base = buffer,
+		.iov_len = arguments->transfer_size
+	};
+
+	ctxinfo->tagged = (struct fi_msg_tagged) {
+		.msg_iov = &ctxinfo->iov,
+		.iov_count = 1,
+		.desc = desc,
+		.addr = rx_address,
+		.tag = tag,
+		.context = &ctxinfo->fi_context,
+		.data = 0
+	};
+
+	if (arguments->use_workqueue && flags & FI_TRIGGER) {
+		struct fi_deferred_work *work = &ctxinfo->def_work;
+		work->triggering_cntr =
+				ctxinfo->fi_trig_context.trigger.threshold.cntr;
+		work->threshold = ctxinfo->fi_trig_context.trigger.threshold.threshold;
+		work->completion_cntr = op_context->tx_cntr;
+
+		flags = flags & ~FI_TRIGGER;
+
+		ctxinfo->tagged_op = (struct fi_op_tagged) {
+			.ep = endpoint,
+			.msg = ctxinfo->tagged,
+			.flags = flags | FI_COMPLETION
+		};
+		work->op_type = FI_OP_TSEND;
+		work->op.tagged = &ctxinfo->tagged_op;
+
+		ret = fi_control(&op_context->domain->fid, FI_QUEUE_WORK, work);
+	} else {
+		ret = fi_tsendmsg(endpoint, &ctxinfo->tagged, flags);
+	}
+
+	return ret;
+}
```

Comment:
I would merge all test files into one file, say test.c.  We have rma, atomic, msg, and tagged tests available.